### PR TITLE
[Product Refactor] Shipping Category

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Part of #9069. Migrates `shipping_category` from Product to Variant.

#### What should we test?

Setting and using shipping category on variants instead of products.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/controllers/spree/admin/variants_controller.rb
Comment (abdellani) [inline app/controllers/spree/admin/variants_controller.rb:35]: Please correct me if I'm wrong.
This will be used to select a default Shipping category when the user wants to create a new variant, right?

What if the master variant refers to a shipping category other than the default one? will the new variant refer to the default shipping category?

Comment (Matt-Yorkley) [reply app/controllers/spree/admin/variants_controller.rb:35]: Master variants are gone :smile: 

This is on the page where the new variant form is shown. Setting this here ensures there won't be a `nil` value for it, and sets the dropdown on the page but a different one can be selected before the form is submitted.

File: app/models/spree/variant.rb
Comment (dacook) [inline app/models/spree/variant.rb:66]: This won't be necessary once https://github.com/openfoodfoundation/openfoodnetwork/pull/11297/files#diff-833d4905682446957f9fc24c1b8d02bf9e1cb594484989aa9d8c921a84aab69e is merged.

I think it's worth waiting until that is merged, then we can remove this line. 

Alternatively in the meantime, it would make it easier to later identify the unnecessary rule by updating the relationship like so:
```
belongs_to :shipping_category, class_name: 'Spree::ShippingCategory', optional: false
```

File: app/views/spree/admin/variants/_form.html.haml
Comment (dacook) [inline app/views/spree/admin/variants/_form.html.haml:74]: We've missed out on `f.error_message_on :shipping_category`; was this not necessary for this form?

File: app/controllers/spree/admin/variants_controller.rb
Comment (dacook) [reply app/controllers/spree/admin/variants_controller.rb:35]: I was thinking it might be helpful to default to a sibling variant's shipping method. But we plan on changing the data model further again so there's not much point doing that now.

File: app/models/spree/variant.rb
Comment (dacook) [inline app/models/spree/variant.rb:262]: I'm wondering what the need for this is? 

If shipping category was required on products before, then they should all be set. Of course, maybe there's some old records that didn't have one somehow, and that problem would be migrated from the product to variant. But we didn't need to handle the problem before, so why now?

File: engines/dfc_provider/app/services/supplied_product_builder.rb
Comment (dacook) [inline engines/dfc_provider/app/services/supplied_product_builder.rb:25]: Maybe this is why we needed to "Set default shipping_category if absent".
I think we could have instead set `shipping_category_id` to pass through to the standard variant. 

But I agree with this commit – the system has a `DefaultShippingCategory`, so we don't need to tell it what category to use.

**Review by dacook — APPROVED** [review]
Looks good, except for the linter error, which I think is best to fix after #11297 has been merged.

File: app/models/spree/variant.rb
Comment (mkllnk) [inline app/models/spree/variant.rb:81]: I'm wondering if the attributes API is better for default values:
```suggestion
    attribute :shipping_category, default: -> { DefaultShippingCategory.find_or_create }
```

File: app/models/spree/variant.rb
Comment (mkllnk) [inline app/models/spree/variant.rb:32]:
```suggestion
    belongs_to :shipping_category, class_name: 'Spree::ShippingCategory', optional: false
```

Instead of the presence validation.

**Review by mkllnk — APPROVED** [review]
Great work.

File: app/models/spree/variant.rb
Comment (mkllnk) [reply app/models/spree/variant.rb:66]: My mentioned PR is keeping the old default for this class. So the `optional: false` is the best solution for now but I don't want that to halt this pull request.

File: app/controllers/spree/admin/products_controller.rb
Comment (dacook) [reply app/controllers/spree/admin/products_controller.rb:20]: 👍 I had to investigate to find this out too. A code comment would have been worthwhile

..so here it is ;)

```suggestion
        # This id is in the form when creating a new product, and is stored on the standard variant that's created with the new product.
        @object.shipping_category_id = DefaultShippingCategory.find_or_create.id
```

File: app/models/spree/variant.rb
Comment (Matt-Yorkley) [reply app/models/spree/variant.rb:81]: Hmm interesting, that might mean we can drop some of the other code in the controller too. I'll take a look.

File: app/views/spree/admin/variants/_form.html.haml
Comment (Matt-Yorkley) [reply app/views/spree/admin/variants/_form.html.haml:74]: No, the two forms work slightly differently.

File: app/models/spree/variant.rb
Comment (Matt-Yorkley) [reply app/models/spree/variant.rb:80]: Nope, this makes the entire test suite explode :frowning_face:

**Review by mkllnk — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @Matt-Yorkley,
Thanks for this one!

### Before staging
- Shipping category is set in product details (on product level):
image
- There is no information on shipping category on the variant's page:
image
- All shipping methods are available at checkout, independently from the shipping category:
image
- I couldn't find any place where the shipping category is displayed. Neither for the customer, nor for the enterprise user (reports etc.)

### After staging
- Shipping category is removed from product details: :heavy_check_mark: 
image
- Shipping category can be set on the variant's page: :heavy_check_mark: 
image
- Creating new shipping categories as super admin seems to work fine:
image

I found two issues though:
1. After staging the PR all variants had the shipping category 'Default' - even if it was set to something else before staging. :x:
2. Saving the shipping category seems successful, but the value isn't actually kept. :x:

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/9dd16e53-8023-4f09-8beb-b8918074b584

### Conclusion
I think we have to move this back to In Dev.

Thanks!

**Comment by Matt-Yorkley** [comment]
Rebased and updated, it should be fixed now :+1:

**Comment by drummer83** [comment]
Hi @Matt-Yorkley,
Thanks for fixing these.

I checked my test cases from above again and also the two open issues:
1. After staging the PR all variants had the shipping category 'Default' - even if it was set to something else before staging. ✔️ 
2. Saving the shipping category seems successful, but the value isn't actually kept. ✔️ 

Previously set shipping categories are now kept and changing them is working fine as well. ✔️ 

I've also tested this quickly on staging FR, where this PR hadn't been staged before. Looking good. ✔️ 

Well done! Thanks! 🙏 
Merging! 🚀

**Comment by drummer83** [comment]
Oh no, now I have merged another PR in between, which caused a conflict.
Sorry, could you please have a look at that, @Matt-Yorkley? Thanks!!
I'll move to In Dev until resolved.

**Comment by dacook** [comment]
Looks like test has passed before conflict was resolved, so it's probably ready to go. @Matt-Yorkley can you confirm if the conflict was inconsequential and doesn't need any further testing? 
If so, let's merge.

**Comment by abdellani** [comment]
@Matt-Yorkley 
Look like there's an issue

https://github.com/openfoodfoundation/openfoodnetwork/assets/16113997/dfc4f2ae-10b1-403b-aaec-c95dbd4cae13

**Pull request merged**
